File: bot-api/java/src/test/java/dev/robocode/tankroyale/botapi/AbstractBotTest.java
package dev.robocode.tankroyale.botapi;

import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import test_utils.MockedServer;

import static org.assertj.core.api.AssertionsForClassTypes.assertThat;

abstract class AbstractBotTest {

    protected MockedServer server;

    protected static final BotInfo botInfo = BotInfo.builder()
            .setName("TestBot")
            .setVersion("1.0")
            .addAuthor("Author 1")
            .addAuthor("Author 2")
            .setDescription("Short description")
            .setHomepage("https://testbot.robocode.dev")
            .addCountryCode("gb")
            .addCountryCode("us")
            .addGameType("classic")
            .addGameType("melee")
            .addGameType("1v1")
            .setPlatform("JVM 19")
            .setProgrammingLang("Java 19")
            .setInitialPosition(InitialPosition.fromString("10, 20, 30"))
            .build();

    protected static class TestBot extends BaseBot {
        TestBot() {
            super(botInfo, MockedServer.getServerUrl());
        }
    }

    @BeforeEach
    void setUp() {
        server = new MockedServer();
        server.start();
    }

    @AfterEach
    void tearDown() {
        server.stop();
    }

    protected static BaseBot start() {
        var bot = new TestBot();
        startAsync(bot);
        return bot;
    }

    protected static void startAsync(BaseBot bot) {
        new Thread(bot::start).start();
    }

    protected static void goAsync(BaseBot bot) {
        new Thread(bot::go).start();
    }

    protected BaseBot startAndAwaitHandshake() {
        var bot = start();
        awaitBotHandshake();
        return bot;
    }

    protected BaseBot startAndAwaitTick() {
        var bot = start();
        awaitTick();
        return bot;
    }

    protected BaseBot startAndAwaitGameStarted() {
        var bot = start();
        awaitGameStarted(bot);
        return bot;
    }

    protected void awaitBotHandshake() {
        assertThat(server.awaitBotHandshake(1000)).isTrue();
    }

    protected void awaitGameStarted(BaseBot bot) {
        assertThat(server.awaitGameStarted(1000)).isTrue();

        boolean noException = false;
        do {
            try {
                bot.getEnergy();
                noException = true;
            } catch (BotException ex) {
                Thread.yield();
            }
        } while (!noException);
    }

    protected void awaitTick() {
        sleep(); // must be processed within the bot api first
        assertThat(server.awaitTick(1000)).isTrue();
    }

    protected void awaitBotIntent() {
        assertThat(server.awaitBotIntent(1000)).isTrue();
    }

    private void sleep() {
        try {
            Thread.sleep(500);
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        }
    }

    protected static boolean exceptionContainsEnvVarName(BotException botException, String envVarName) {
        return botException.getMessage().toUpperCase().contains(envVarName);
    }
}

File: bot-api/dotnet/Robocode.TankRoyale.BotApi.Tests/src/AbstractBotTest.cs
using System;
using System.Threading;
using System.Threading.Tasks;
using NUnit.Framework;
using Robocode.TankRoyale.BotApi.Tests.Test_utils;

namespace Robocode.TankRoyale.BotApi.Tests;

public class AbstractBotTest
{
    protected MockedServer Server;

    protected static readonly BotInfo BotInfo = BotInfo.Builder()
        .SetName("TestBot")
        .SetVersion("1.0")
        .AddAuthor("Author 1")
        .AddAuthor("Author 2")
        .SetDescription("Short description")
        .SetHomepage("https://testbot.robocode.dev")
        .AddCountryCode("gb")
        .AddCountryCode("us")
        .AddGameType("classic")
        .AddGameType("1v1")
        .AddGameType("melee")
        .SetPlatform(".Net 6")
        .SetProgrammingLang("C# 10")
        .SetInitialPosition(InitialPosition.FromString("10, 20, 30"))
        .Build();
    
    private class TestBot : BaseBot {
        public TestBot() : base(BotInfo, (new Uri("ws://127.0.0.1:" + MockedServer.Port)))
        {
        }
    }
    
    [SetUp]
    public void SetUp()
    {
        Server = new MockedServer();
        Server.Start();
    }

    [TearDown]
    public void Teardown()
    {
        Server.Stop();
    }
    
    protected static BaseBot Start()
    {
        var bot = new TestBot();
        StartAsync(bot);
        return bot;
    }

    protected static void StartAsync(BaseBot bot)
    {
        Task.Run(bot.Start);
    }

    protected static void GoAsync(BaseBot bot)
    {
        Task.Run(bot.Go);
    }

    protected BaseBot StartAndAwaitHandshake()
    {
        var bot = Start();
        AwaitBotHandshake();
        return bot;
    }

    protected BaseBot StartAndAwaitTick()
    {
        var bot = Start();
        AwaitTick();
        return bot;
    }

    protected BaseBot StartAndAwaitGameStarted()
    {
        var bot = Start();
        AwaitGameStarted();
        return bot;
    }

    protected void AwaitBotHandshake()
    {
        Assert.That(Server.AwaitBotHandshake(2000), Is.True);
    }

    private void AwaitGameStarted()
    {
        Assert.That(Server.AwaitGameStarted(1000), Is.True);
        Sleep(); // must be processed within the bot api first
    }

    protected void AwaitTick()
    {
        Assert.That(Server.AwaitTick(1000), Is.True);
        Sleep(); // must be processed within the bot api first
    }

    protected void AwaitBotIntent()
    {
        Assert.That(Server.AwaitBotIntent(1000), Is.True);
    }

    private static void Sleep()
    {
        Thread.Sleep(500);
    }
    
    protected static bool ExceptionContainsEnvVarName(BotException botException, string envVarName) =>
        botException != null && botException.Message.ToUpper().Contains(envVarName);
}